### PR TITLE
[shardformer/sequence parallel] Overlap input gather and grad computation during col backward

### LINKED ISSUE #4400: [shardformer/sequence parallel] support overlap gather input and gradient computation during col backward

### PR DESCRIPTION
**Pull request by FoolPlayer**
## 📌 Checklist before creating the PR

- [x] I have created an issue for this PR for traceability
- [x] The title follows the standard format: `[doc/gemini/tensor/...]: A concise description`
- [x] I have added relevant tags if possible for us to better distinguish different PRs

## 🚨 Issue number

> Link this PR to your issue with words like fixed to automatically close the linked issue upon merge
>
> e.g. `fixed #1234`, `closed #1234`, `resolved #1234`
fixed #4400 

## 📝 What does this PR do?

> Summarize your work here.
> if you have any plots/diagrams/screenshots/tables, please attach them here.
Overlap input gather and grad computation during col backward, support linear so far

## 💥 Checklist before requesting a review

- [x] I have linked my PR to an issue ([instruction](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue))
- [x] My issue clearly describes the problem/feature/proposal, with diagrams/charts/table/code if possible
- [x] I have performed a self-review of my code
- [x] I have added thorough tests.
- [x] I have added docstrings for all the functions/methods I implemented

## ⭐️ Do you enjoy contributing to Colossal-AI?

- [x] 🌝 Yes, I do.
- [ ] 🌚 No, I don't.

Tell us more if you don't enjoy contributing to Colossal-AI.

### CODE REVIEW

File: colossalai/shardformer/layer/_operation.py [hunk 1]
```diff
@@ -175,42 +177,81 @@ def backward(ctx, grad_output):
...
+            reducescatter_handle = dist.reduce_scatter(output, input_list, group=process_group, async_op=True)
+            with torch.cuda.stream(calculate_stream):
+                input_parallel = torch.cat(tensor_list, dim=dim).contiguous()
+                if len(input_parallel.shape) > 2:
+                    input_parallel = input_parallel.view(-1, input_parallel.shape[-1])
+                print(grad_output.shape, input_parallel.shape)
+                grad_weight = grad_output.t().matmul(input_parallel)
+
+            reducescatter_handle.wait()
+
```
Comment (ver217) [inline colossalai/shardformer/layer/_operation.py:253]: For safety, you'd better wait computation stream here.

Comment (FoolPlayer) [reply colossalai/shardformer/layer/_operation.py:252]: change this line as `torch.cuda.synchronize()`

**Review by ver217 — APPROVED** [review]
